sys/shell/shell.c

```diff
@@ -92,6 +98,23 @@ static shell_command_handler_t search_commands(const shell_command_t *entry,
     return NULL;
 }
 
+static shell_command_handler_t search_commands_xfa(char *command)
+{
+#if IS_USED(MODULE_SHELL_COMMAND_XFA)
+    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
+
+    for (unsigned i = 0; i < n; i++) {
+        const volatile shell_command_t *entry = shell_commands_xfa[i];
```

Comment:
Why `volatile` here?

Reply:
Hmm, sounds a bit like a foot gun, as it is easy to forget.

Also, this marks the memory `entry` points to as `volatile`. So very access to a member of `entry` will result in a read from memory, rather than accessing a register that is still holding the value. This is not an issue here with strings, but it is not too hard to imagine situations in which this would result in poorly optimized code.

I think that it might be that the compiler only falsely assumes that the length is zero. There are some bugs in GCC leading to unoptimised code surrounding `volatile` accesses. If this is the case, newer versions of GCC without this bug would still optimize out the loop. I will try to find some time to dig into this and more specifically tell the compiler where not to optimize.

Reply:
> Hmm, sounds a bit like a foot gun, as it is easy to forget.

what would be the footgun? gcc does assume the array has length 0 (it is defined as such ...).
The volatile is enforced by the compiler's type check (and by the XFA macros), it should not be possible to accidentally not use volatilr.

The missed optimization opportunities are a thing. I'd hope they's not be so bad...

Reply:
> The volatile is enforced by the compiler's type check (and by the XFA macros), it should not be possible to accidentally not use volatilr.

OK, no footgun then :-)